File: src/states/game_over_state.py
import pygame
from src import config
from src.engine.state_machine import State
from src.engine.ui import Button
from src.gameplay.run_manager import RunManager

class GameOverState(State):
    def __init__(self, state_machine, run_manager):
        super().__init__(state_machine, run_manager)
        self.buttons = []
        self.title_font = config.get_font("typewriter", 56)
        self.lbl_font = config.get_font("sans", 24)
        self.desc_font = config.get_font("sans", 18)
        self.stat_font = config.get_font("sans", 22)
        
        # Result type: "fired" (lose) or "published" (win)
        self.result = "fired"

    def enter(self, **kwargs):
        self.result = kwargs.get("result", "fired")
        self.buttons.clear()
        
        # Action buttons
        self.buttons.append(Button(
            x=config.SCREEN_WIDTH // 2 - 220,
            y=460,
            width=200,
            height=50,
            text="Try Again",
            callback=self.retry_run,
            color=(46, 180, 110) # Green
        ))
        self.buttons.append(Button(
            x=config.SCREEN_WIDTH // 2 + 20,
            y=460,
            width=200,
            height=50,
            text="Main Menu",
            callback=self.goto_menu,
            color=(52, 152, 219) # Blue
        ))

    def retry_run(self):
        # Reset run manager to brand new state
        # (Re-instantiating RunManager is the easiest way to reset the whole state)
        new_rm = RunManager()
        self.state_machine.run_manager = new_rm
        
        # Re-link states to new run manager
        for s in self.state_machine.states.values():
            s.run_manager = new_rm
            
        self.state_machine.change_state("blind_select")

    def goto_menu(self):
        self.state_machine.change_state("menu")

    def handle_events(self, events):
        mpos = pygame.mouse.get_pos()
        for btn in self.buttons:
            btn.check_hover(mpos)
            
        for event in events:
            for btn in self.buttons:
                if btn.handle_event(event, mpos):
                    break

    def update(self, dt):
        for btn in self.buttons:
            btn.update(dt)

    def draw(self, surface):
        surface.fill(config.COLOR_DESK)
        
        # Draw paper background
        paper_rect = pygame.Rect(config.SCREEN_WIDTH // 2 - 320, 60, 640, 600)
        pygame.draw.rect(surface, (15, 15, 20), paper_rect.inflate(8, 8), border_radius=8)
        pygame.draw.rect(surface, config.COLOR_PAPER, paper_rect, border_radius=6)
        
        # Draw stamp header
        if self.result == "published":
            title_text = "PUBLISHED!"
            accent_col = config.COLOR_CLUE_GREEN
            desc_text = "Congratulations! Your manuscript has become a national bestseller."
        else:
            title_text = "YOU ARE FIRED"
            accent_col = config.COLOR_ACCENT
            desc_text = "You ran out of submissions before meeting the Hype requirements."
            
        title_surf = self.title_font.render(title_text, True, accent_col)
        title_rect = title_surf.get_rect(center=(config.SCREEN_WIDTH // 2, 140))
        surface.blit(title_surf, title_rect)
        
        # Subtitle
        desc_surf = self.desc_font.render(desc_text, True, config.COLOR_TEXT_DARK)
        desc_rect = desc_surf.get_rect(center=(config.SCREEN_WIDTH // 2, 210))
        surface.blit(desc_surf, desc_rect)
        
        pygame.draw.line(surface, config.COLOR_TEXT_MUTED, (config.SCREEN_WIDTH // 2 - 250, 240), (config.SCREEN_WIDTH // 2 + 250, 240), 1)
        
        # Draw stats list
        stats = [
            f"Chapters Completed: {self.run_manager.chapter - 1} / 8",
            f"Final Royalties Earned: ${self.run_manager.royalties}",
            f"Active Tropes Equipped: {len(self.run_manager.tropes)}",
        ]
        
        y_off = 265
        for stat in stats:
            s_surf = self.stat_font.render(stat, True, config.COLOR_TEXT_DARK)
            s_rect = s_surf.get_rect(center=(config.SCREEN_WIDTH // 2, y_off))
            surface.blit(s_surf, s_rect)
            y_off += 30
            
        # Draw Tropes final inventory
        tropes_lbl = self.lbl_font.render("Your Desk Tropes:", True, config.COLOR_TEXT_DARK)
        surface.blit(tropes_lbl, (config.SCREEN_WIDTH // 2 - 240, 360))
        
        if not self.run_manager.tropes:
            none_surf = self.desc_font.render("None (No modifiers equipped)", True, config.COLOR_TEXT_MUTED)
            surface.blit(none_surf, (config.SCREEN_WIDTH // 2 - 240, 395))
        else:
            y_t = 395
            for trope in self.run_manager.tropes[:3]: # draw up to 3 tropes
                t_surf = self.desc_font.render(f"- {trope.name}: {trope.description[:55]}...", True, config.COLOR_TEXT_DARK)
                surface.blit(t_surf, (config.SCREEN_WIDTH // 2 - 240, y_t))
                y_t += 22

        # Draw buttons
        for btn in self.buttons:
            btn.draw(surface)

File: love2d/src/states/game_over_state.lua
local config = require("src/config")
local ui = require("src/engine/ui")
local State = require("src/engine/state_machine").State

local GameOverState = setmetatable({}, State)
GameOverState.__index = GameOverState

function GameOverState.new(state_machine, run_manager)
    local self = State.new(state_machine, run_manager)
    setmetatable(self, GameOverState)
    
    self.buttons = {}
    self.title_font = config.get_font("typewriter", 56)
    self.lbl_font = config.get_font("sans", 24)
    self.desc_font = config.get_font("sans", 18)
    self.stat_font = config.get_font("sans", 22)
    
    self.result = "fired"
    
    return self
end

function GameOverState:enter(kwargs)
    self.result = kwargs.result or "fired"
    self.buttons = {}
    
    table.insert(self.buttons, ui.Button.new(
        config.SCREEN_WIDTH / 2 - 220,
        460,
        200,
        50,
        "Try Again",
        function() self:retry_run() end,
        {46/255, 180/255, 110/255}
    ))
    table.insert(self.buttons, ui.Button.new(
        config.SCREEN_WIDTH / 2 + 20,
        460,
        200,
        50,
        "Main Menu",
        function() self:goto_menu() end,
        {52/255, 152/255, 219/255}
    ))
end

function GameOverState:retry_run()
    local RunManager = require("src/gameplay/run_manager")
    local new_rm = RunManager.new()
    self.state_machine.run_manager = new_rm
    
    for _, s in pairs(self.state_machine.states) do
        s.run_manager = new_rm
    end
    
    self.state_machine:change_state("assignment_select")
end

function GameOverState:goto_menu()
    self.state_machine:change_state("menu")
end

function GameOverState:update(dt)
    local mx, my = love.mouse.getPosition()
    mx, my = config.mx or mx, config.my or my
    
    for _, btn in ipairs(self.buttons) do
        btn:check_hover(mx, my)
        btn:update(dt)
    end
end

function GameOverState:mousepressed(x, y, button, istouch, presses)
    local mx, my = config.mx or x, config.my or y
    for _, btn in ipairs(self.buttons) do
        if btn:handle_event("mousepressed", mx, my, button) then
            break
        end
    end
end

function GameOverState:draw()
    ui.draw_background()
    
    local px, py, pw, ph = config.SCREEN_WIDTH / 2 - 320, 60, 640, 600
    love.graphics.setColor(15/255, 15/255, 20/255, 1.0)
    love.graphics.rectangle("fill", px - 8, py - 8, pw + 16, ph + 16, 8, 8)
    
    if config.images.overlay_paper then
        love.graphics.setColor(1.0, 1.0, 1.0, 1.0)
        love.graphics.draw(config.images.overlay_paper, px, py, 0, pw / config.images.overlay_paper:getWidth(), ph / config.images.overlay_paper:getHeight())
    else
        love.graphics.setColor(config.COLOR_PAPER[1], config.COLOR_PAPER[2], config.COLOR_PAPER[3], 1.0)
        love.graphics.rectangle("fill", px, py, pw, ph, 6, 6)
    end
    
    local title_text, accent_col, desc_text
    if self.result == "published" then
        title_text = "PUBLISHED!"
        accent_col = config.COLOR_CLUE_GREEN
        desc_text = "Congratulations! Your manuscript has become a national bestseller."
    else
        title_text = "YOU ARE FIRED"
        accent_col = config.COLOR_ACCENT
        desc_text = "You ran out of submissions before meeting the Hype requirements."
    end
    
    love.graphics.setFont(self.title_font)
    love.graphics.setColor(accent_col[1], accent_col[2], accent_col[3], 1.0)
    local tw = self.title_font:getWidth(title_text)
    love.graphics.print(title_text, config.SCREEN_WIDTH / 2 - tw / 2, 115)
    
    love.graphics.setFont(self.desc_font)
    love.graphics.setColor(config.COLOR_TEXT_DARK[1], config.COLOR_TEXT_DARK[2], config.COLOR_TEXT_DARK[3], 1.0)
    local dw = self.desc_font:getWidth(desc_text)
    love.graphics.print(desc_text, config.SCREEN_WIDTH / 2 - dw / 2, 200)
    
    love.graphics.setColor(config.COLOR_TEXT_MUTED[1], config.COLOR_TEXT_MUTED[2], config.COLOR_TEXT_MUTED[3], 1.0)
    love.graphics.line(config.SCREEN_WIDTH / 2 - 250, 240, config.SCREEN_WIDTH / 2 + 250, 240)
    
    local stats = {
        "Chapters Completed: " .. (self.run_manager.chapter - 1) .. " / 8",
        "Final Royalties Earned: $" .. self.run_manager.royalties,
        "Active Tropes Equipped: " .. #self.run_manager.tropes,
    }
    
    local y_off = 265
    love.graphics.setFont(self.stat_font)
    love.graphics.setColor(config.COLOR_TEXT_DARK[1], config.COLOR_TEXT_DARK[2], config.COLOR_TEXT_DARK[3], 1.0)
    for _, stat in ipairs(stats) do
        local sw = self.stat_font:getWidth(stat)
        love.graphics.print(stat, config.SCREEN_WIDTH / 2 - sw / 2, y_off)
        y_off = y_off + 30
    end
    
    love.graphics.setFont(self.lbl_font)
    love.graphics.setColor(config.COLOR_TEXT_DARK[1], config.COLOR_TEXT_DARK[2], config.COLOR_TEXT_DARK[3], 1.0)
    love.graphics.print("Your Desk Tropes:", config.SCREEN_WIDTH / 2 - 240, 360)
    
    love.graphics.setFont(self.desc_font)
    if #self.run_manager.tropes == 0 then
        love.graphics.setColor(config.COLOR_TEXT_MUTED[1], config.COLOR_TEXT_MUTED[2], config.COLOR_TEXT_MUTED[3], 1.0)
        love.graphics.print("None (No modifiers equipped)", config.SCREEN_WIDTH / 2 - 240, 395)
    else
        local y_t = 395
        love.graphics.setColor(config.COLOR_TEXT_DARK[1], config.COLOR_TEXT_DARK[2], config.COLOR_TEXT_DARK[3], 1.0)
        for idx = 1, math.min(3, #self.run_manager.tropes) do
            local trope = self.run_manager.tropes[idx]
            local desc_sub = trope.description:sub(1, 55)
            if #trope.description > 55 then
                desc_sub = desc_sub .. "..."
            end
            love.graphics.print("- " .. trope.name .. ": " .. desc_sub, config.SCREEN_WIDTH / 2 - 240, y_t)
            y_t = y_t + 22
        end
    end
    
    for _, btn in ipairs(self.buttons) do
        btn:draw()
    end
end

return GameOverState
